File: problems/lru-cache/lru_cache.c

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int key;
  int val;
  struct Node *prev;
  struct Node *next;
} Node;

typedef struct LRUCache {
  int capacity;
  int count;
  Node *head;
  Node *tail;
  Node *map[10000];
} LRUCache;

Node *node_create(int key, int val) {
  Node *node = (Node *)malloc(sizeof(Node));
  node->key = key;
  node->val = val;
  node->prev = NULL;
  node->next = NULL;
  return node;
}

LRUCache *cache_create(int capacity) {
  LRUCache *cache = (LRUCache *)malloc(sizeof(LRUCache));
  cache->capacity = capacity;
  cache->count = 0;
  cache->head = node_create(0, 0);
  cache->tail = node_create(0, 0);
  cache->head->next = cache->tail;
  cache->tail->prev = cache->head;

  for (int i = 0; i < 10000; i++) {
    cache->map[i] = NULL;
  }
  return cache;
}

void remove_node(Node *node) {
  node->prev->next = node->next;
  node->next->prev = node->prev;
}

void add_to_head(LRUCache *cache, Node *node) {
  node->prev = cache->head;
  node->next = cache->head->next;
  cache->head->next->prev = node;
  cache->head->next = node;
}

int get(LRUCache *cache, int key) {
  Node *node = cache->map[key];
  if (node) {
    remove_node(node);
    add_to_head(cache, node);
    return node->val;
  }
  return -1;
}

void put(LRUCache *cache, int key, int val) {
  Node *node = cache->map[key];
  if (node) {
    node->val = val;
    remove_node(node);
    add_to_head(cache, node);
  } else {
    Node *new_node = node_create(key, val);
    cache->map[key] = new_node;
    add_to_head(cache, new_node);
    cache->count++;
    if (cache->count > cache->capacity) {
      Node *lru = cache->tail->prev;
      remove_node(lru);
      cache->map[lru->key] = NULL;
      cache->count--;
      // free(lru); Leaked to match Salt
    }
  }
}

int main() {
  int checksum = 0;
  for (int k = 0; k < 1000; k++) {
    LRUCache *cache = cache_create(100);
    for (int i = 0; i < 1000; i++) {
      put(cache, i, i * 10);
      if (i % 2 == 0) {
        checksum += get(cache, i / 2);
      }
    }
  }
  printf("Checksum: %d\n", checksum);
  return 0;
}

File: problems/lru-cache/lru_cache.rs

use std::ptr;

struct Node {
    key: i32,
    val: i32,
    prev: *mut Node,
    next: *mut Node,
}

struct LRUCache {
    capacity: i32,
    count: i32,
    head: *mut Node,
    tail: *mut Node,
    map: Vec<*mut Node>,
}

impl LRUCache {
    fn new(capacity: i32) -> Self {
        unsafe {
            let head = Box::into_raw(Box::new(Node { key: 0, val: 0, prev: ptr::null_mut(), next: ptr::null_mut() }));
            let tail = Box::into_raw(Box::new(Node { key: 0, val: 0, prev: ptr::null_mut(), next: ptr::null_mut() }));
            (*head).next = tail;
            (*tail).prev = head;
            
            LRUCache {
                capacity,
                count: 0,
                head,
                tail,
                map: vec![ptr::null_mut(); 10000],
            }
        }
    }
    
    fn remove_node(&mut self, node: *mut Node) {
        unsafe {
            let prev = (*node).prev;
            let next = (*node).next;
            (*prev).next = next;
            (*next).prev = prev;
        }
    }
    
    fn add_to_head(&mut self, node: *mut Node) {
        unsafe {
            (*node).prev = self.head;
            (*node).next = (*self.head).next;
            (*(*self.head).next).prev = node;
            (*self.head).next = node;
        }
    }
    
    fn get(&mut self, key: i32) -> i32 {
        let node = self.map[key as usize];
        if !node.is_null() {
            self.remove_node(node);
            self.add_to_head(node);
            unsafe { return (*node).val; }
        }
        -1
    }
    
    fn put(&mut self, key: i32, val: i32) {
        let node = self.map[key as usize];
        unsafe {
            if !node.is_null() {
                (*node).val = val;
                self.remove_node(node);
                self.add_to_head(node);
            } else {
                let new_node = Box::into_raw(Box::new(Node { key, val, prev: ptr::null_mut(), next: ptr::null_mut() }));
                self.map[key as usize] = new_node;
                self.add_to_head(new_node);
                self.count += 1;
                if self.count > self.capacity {
                    let lru = (*self.tail).prev;
                    self.remove_node(lru);
                    self.map[(*lru).key as usize] = ptr::null_mut();
                    self.count -= 1;
                    // Box::from_raw(lru); // Leak to match others
                }
            }
        }
    }
}

fn main() {
    let mut checksum = 0;
    for _ in 0..1000 {
        let mut cache = LRUCache::new(100);
        for i in 0..1000 {
            cache.put(i, i * 10);
            if i % 2 == 0 {
                checksum += cache.get(i / 2);
            }
        }
    }
    println!("Checksum: {}", checksum);
    // std::hint::black_box(checksum);
}
